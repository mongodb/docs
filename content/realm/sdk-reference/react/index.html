<!doctype html>
<head>
  <meta http-equiv="refresh" content="0; url=https://mongodb.com/docs/realm-sdks/react/latest" />
</head>
<body>
  <p><a href="https://mongodb.com/docs/realm-sdks/react/latest">Redirect</a></p>
</body>
