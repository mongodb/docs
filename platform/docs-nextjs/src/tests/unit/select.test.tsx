import { useState } from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Select, { type SelectOption } from '@/components/select';

const DEFAULT_ATLAS_CHOICE = { text: 'MongoDB Atlas', value: 'atlas' };
const DEFAULT_SERVER_CHOICE = { text: 'MongoDB Server', value: 'server' };
const DEFAULT_CHOICES = [DEFAULT_SERVER_CHOICE, DEFAULT_ATLAS_CHOICE];

type SelectControllerProps = {
  choices?: SelectOption[];
  customOnChange?: ({ value }: { value: string }) => void;
  defaultText?: string;
  disabled?: boolean;
  label?: string;
  value?: string;
};
// Simple wrapper to add state control
const SelectController = ({
  choices = DEFAULT_CHOICES,
  customOnChange,
  defaultText = '',
  disabled = false,
  label,
  value,
}: SelectControllerProps) => {
  const [selectValue, setSelectValue] = useState(value);
  const onChange = ({ value }: { value: string }) => {
    if (customOnChange) customOnChange({ value });
    setSelectValue(value);
  };
  return (
    <Select
      choices={choices}
      onChange={onChange}
      defaultText={defaultText}
      disabled={disabled}
      label={label}
      value={selectValue}
    />
  );
};

describe('Select', () => {
  // Helper to open the dropdown passed a series of simulate args (click, keypress)
  const dropdownOpen = async (props = {}, openWithKeyboard = false) => {
    const user = userEvent.setup();
    const wrapper = render(<SelectController {...props} />);
    // Dropdown should be closed by default
    const dropdown = wrapper.container.querySelector('button');
    expect(dropdown).toHaveAttribute('aria-expanded', 'false');
    if (openWithKeyboard) {
      await user.tab();
      await user.keyboard('{Enter}');
    } else {
      await user.click(dropdown as Element);
    }
    // Now it should be open
    expect(dropdown).toHaveAttribute('aria-expanded', 'true');
    return wrapper;
  };

  it('renders select correctly', () => {
    const wrapper = render(<SelectController />);
    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it('displays default text', () => {
    const defaultText = 'Some default texts';
    const wrapper = render(<SelectController defaultText={defaultText} />);
    const renderedText = wrapper.getByText(defaultText);
    expect(renderedText).toBeTruthy();
  });

  it('conditionally should render a label', () => {
    const labelText = 'Select Label';
    const wrapperWithLabel = render(<SelectController label={labelText} />);
    expect(wrapperWithLabel.getByText(labelText)).toBeTruthy();
  });

  it('opens a dropdown with options when clicked', async () => {
    await dropdownOpen();
  });

  it('opens a dropdown with options with the enter key for accessibility', async () => {
    await dropdownOpen({ key: 'Enter' }, true);
  });

  it('passes disabled prop through to select implementation when given', () => {
    const wrapper = render(<SelectController disabled />);
    expect(wrapper.container.querySelector('button')).toHaveAttribute('aria-disabled', 'true');
  });

  it('closes the dropdown by clicking again on the toggle parent', async () => {
    const user = userEvent.setup();
    const wrapper = await dropdownOpen();
    const dropdownButton = wrapper.container.querySelector('button');
    await user.click(dropdownButton as Element);
    // Dropdown was previously open, it should now be closed
    expect(dropdownButton).toHaveAttribute('aria-expanded', 'false');
  });

  it('updates the selected text when an item is clicked', async () => {
    const user = userEvent.setup();
    const defaultText = 'Default Text';
    const wrapper = await dropdownOpen({ defaultText });
    let renderedText = wrapper.container.querySelector('button');
    expect(renderedText?.textContent).toBe(defaultText);
    // Implementation stores the 'select' in the first option field, as rendered html
    const firstOption = wrapper.queryAllByRole('option')[0];
    expect(firstOption.textContent).toBe(DEFAULT_CHOICES[0].text);
    await user.click(firstOption);
    //check the button text to make sure it updated
    renderedText = wrapper.container.querySelector('button');
    expect(renderedText?.textContent).toBe(DEFAULT_CHOICES[0].text);
  });

  it('passes the value attribute of original choice to the onChange callback', async () => {
    const user = userEvent.setup();
    const customOnChange = jest.fn();
    const wrapper = await dropdownOpen({ customOnChange });
    const firstOption = wrapper.queryAllByRole('option')[0];
    expect(firstOption.textContent).toBe(DEFAULT_CHOICES[0].text);
    await user.click(firstOption);
    expect(customOnChange.mock.calls.length).toBe(1);
    expect(customOnChange.mock.calls[0][0]).toStrictEqual({ value: DEFAULT_CHOICES[0].value });
  });

  it('should update selected text given a value', () => {
    const wrapper = render(<SelectController value={DEFAULT_CHOICES[0].value} />);
    expect(wrapper.getByText(DEFAULT_CHOICES[0].text)).toBeTruthy();
  });

  it('should reset the form when null/empty string is passed as value', () => {
    const defaultText = 'Reset text';
    // Bypass SelectController so we can directly rerender with equivalent component + state
    const wrapper = render(
      <Select
        defaultText={defaultText}
        choices={DEFAULT_CHOICES}
        onChange={() => {}}
        value={DEFAULT_CHOICES[0].value}
      />,
    );
    expect(wrapper.getByText(DEFAULT_CHOICES[0].text)).toBeTruthy();
    wrapper.rerender(<Select defaultText={defaultText} choices={DEFAULT_CHOICES} onChange={() => {}} value={''} />);
    expect(wrapper.getByText(defaultText)).toBeTruthy();
  });
});
