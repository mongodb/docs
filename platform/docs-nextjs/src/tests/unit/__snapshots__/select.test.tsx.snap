// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Select renders select correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
}

@media only screen and (max-width: 1024px) {
  .emotion-0 label,
  .emotion-0 p,
  .emotion-0 button,
  .emotion-0 div,
  .emotion-0 span {
    font-size: 13px;
  }
}

.emotion-0 label {
  color: #001E2B;
}

.dark-theme .emotion-0 label {
  color: #E8EDEB;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-2>button {
  background-color: #FFFFFF;
  text-align: left;
}

.emotion-2>button>*:last-child>svg {
  color: #5C6C75;
}

.emotion-2>button:focus-visible {
  box-shadow: 0 0 0 3px #0498EC;
  border-color: rgba(255, 255, 255, 0);
}

.dark-theme .emotion-2>button {
  border-color: #889397;
  background-color: #112733;
}

.dark-theme .emotion-2>button>*:last-child>svg {
  color: #C1C7C6;
}

.dark-theme .emotion-2>button:hover,
.dark-theme .emotion-2>button:active,
.dark-theme .emotion-2>button:focus {
  background-color: #112733;
  color: #E8EDEB;
}

.dark-theme .emotion-2>button:focus-visible {
  background-color: #112733;
  box-shadow: 0 0 0 3px #0498EC;
  border-color: rgba(255, 255, 255, 0);
}

.emotion-3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  padding: 0;
  margin: 0;
  background-color: transparent;
  border: 1px solid transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-transition: all 150ms ease-in-out;
  transition: all 150ms ease-in-out;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  z-index: 0;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  border-radius: 6px;
  background-color: #F9FBFA;
  border-color: #889397;
  color: #001E2B;
  font-size: 13px;
  line-height: 20px;
  font-weight: 500;
  height: 36px;
  font-weight: 400;
  background-color: #FFFFFF;
  color: #889397;
  width: 100%;
  font-size: 13px;
}

.emotion-3:focus-visible,
.emotion-3[data-focus="true"] {
  outline: none;
}

.emotion-3:active,
.emotion-3[data-active="true"],
.emotion-3:focus,
.emotion-3:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-3:focus-visible,
.emotion-3[data-focus="true"] {
  color: #001E2B;
}

.emotion-3:hover,
.emotion-3[data-hover="true"],
.emotion-3:active,
.emotion-3[data-active="true"] {
  color: #001E2B;
  background-color: #FFFFFF;
  box-shadow: 0 0 0 3px #E8EDEB;
}

.emotion-3:focus-visible,
.emotion-3[data-focus="true"] {
  background-color: #FFFFFF;
  box-shadow: 0 0 0 2px #FFFFFF,0 0 0 4px #0498EC;
}

.emotion-3>*:last-child {
  grid-template-columns: 1fr 16px;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
}

.emotion-3>*:last-child>svg {
  justify-self: right;
  width: 16px;
  height: 16px;
}

.emotion-3>*:last-child>svg {
  color: #5C6C75;
}

.emotion-3>*:last-child {
  padding: 0 8px 0 12px;
}

.emotion-3:focus-visible {
  box-shadow: 0 0 0 3px #0498EC;
  border-color: rgba(255, 255, 255, 0);
}

@media only screen and (max-width: 768px) and (hover: none) and (pointer: coarse),@media only screen and (max-width: 768px) and (hover: none) and (pointer: none) {
  .emotion-3 {
    height: 36px;
    font-size: 16px;
  }
}

.emotion-4 {
  overflow: hidden;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  border-radius: 5px;
}

.emotion-5 {
  display: grid;
  grid-auto-flow: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  width: 100%;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  z-index: 0;
  -webkit-transition: all 150 ease-in-out;
  transition: all 150 ease-in-out;
  padding: 0 11px;
  gap: 6px;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  gap: 4px;
  overflow: hidden;
}

.emotion-7 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  max-width: 100%;
}

.emotion-8 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-9 {
  color: #889397;
  height: 16px;
  width: 16px;
  justify-self: left;
}

.emotion-10 {
  display: none;
}

.emotion-11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 4px;
}

<div
    class="undefined emotion-0"
  >
    <div
      class="emotion-1 emotion-2"
      data-lgid="lg-select"
      data-testid="lg-select"
    >
      <button
        aria-controls="select-1-menu"
        aria-describedby="select-1-description"
        aria-disabled="false"
        aria-expanded="false"
        aria-invalid="false"
        aria-labelledby="select"
        class="lg-ui-button-0000 emotion-3"
        data-lgid="lg-select-trigger"
        data-testid="lg-select"
        id="select-2"
        type="button"
        value=""
      >
        <div
          class="emotion-4"
        />
        <div
          class="emotion-5"
        >
          <div
            class="emotion-6"
          >
            <div
              class="lg-ui-select-menu-0000 emotion-7"
              data-lgid="lg-select-button_text"
              data-testid="lg-select-button_text"
            />
          </div>
          <svg
            alt=""
            aria-hidden="true"
            class="emotion-8 emotion-9"
            height="16"
            role="presentation"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M8.67903 10.7962C8.45271 11.0679 8.04729 11.0679 7.82097 10.7962L4.63962 6.97649C4.3213 6.59428 4.5824 6 5.06866 6L11.4313 6C11.9176 6 12.1787 6.59428 11.8604 6.97649L8.67903 10.7962Z"
              fill="currentColor"
            />
          </svg>
        </div>
      </button>
      <span
        class="emotion-10"
      />
      <div
        aria-live="polite"
        aria-relevant="all"
        class="emotion-11"
        data-lgid="lg-form_field-feedback"
        data-testid="lg-form_field-feedback"
      />
    </div>
  </div>
</DocumentFragment>
`;
